### PR TITLE
Ac/pricing

### PR DESCRIPTION
**Pull request by AxithChoudhary**
made changes in the pricing and promotion.
added the preprocessing layer.

### CODE REVIEW

File: server/services/utils/pricing.py [hunk 1]
```diff
@@ -7,52 +7,20 @@
...
+# from services.utils.prompts.pricing_prompts import select_product, price_prompt, promotion_prompt
+# from services.utils.llms import llm_call
  
+from llms import llm_call
+from prompts.pricing_prompts import select_product, price_prompt, promotion_prompt
+import asyncio
 import os
-# from langchain_openai import AzureOpenAI
- 
 dotenv.load_dotenv()
```
Comment (gaurav-9721) [inline server/services/utils/pricing.py:21]: from dotenv import load_dotenv()

Then use load_dotenv() separatly

File: server/services/utils/processing.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    
+    Args:
+        text (str): The input text to be cleaned.
+        
+    Returns:
+        str: The cleaned text with special characters removed.
+    """
+
+    text = re.sub(r'(\d+)([a-zA-Z])', r'\1 \2', text)
+    text = re.sub(r'([a-zA-Z])(\d+)', r'\1 \2', text)
```
Comment (gaurav-9721) [inline server/services/utils/processing.py:32]: store the patterns as variables, give them meaningful name then use in re.sub

File: server/services/utils/processing.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+TOKEN_SIZE = 2800
+
+def remove_special_characters(text:str)->str:
+
+    """
+    Removes special characters from the given text and returns the cleaned text.
+    
+    Args:
+        text (str): The input text to be cleaned.
+
```
Comment (gaurav-9721) [inline server/services/utils/processing.py:26]: Remove Doctstrings

File: server/services/utils/pricing.py [hunk 4]
```diff
@@ -113,151 +109,77 @@ def get_promotion(price, offers):
...
-#     if os.path.exists(r".dummy_data 4.txt"):
-#         with open(r".dummy_data 4.txt", "r") as file:
-#             data = eval(file.read())
-#         print(data)
-   
-#     print(get_price(query, data))
+    return product_offer
+
+if __name__ == "__main__":
+    query = "Men Polo Red T-shirt"
```
Comment (ShubhamVora-1206) [inline server/services/utils/pricing.py:177]: Please remove testing code from PRs

File: server/services/utils/processing.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            (each_slice) * window_size + each_index
+            for each_index in indexes_list
+        ]
+
+    return final_result
+
+
+if __name__ == "__main__":
+
+    query = "iphone 15 pro max black 256GB"
```
Comment (ShubhamVora-1206) [inline server/services/utils/processing.py:105]: Please remove testing code/Hard coded queries from PRs.

File: server/services/utils/prompts/preprocessing_prompt.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+
+preprocessing_prompt = """
+    You are a smart product finder that identifies products in a list based on user queries. make sure you find the right products for the user
+    Given a list of product details, return the index of the product that matches the user query you can return more then one product. it should strictly have all the things that user require.
```
Comment (ShubhamVora-1206) [inline server/services/utils/prompts/preprocessing_prompt.py:4]: spell check: more then one ❌ more than one ✅

File: server/services/utils/processing.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+    Args:
+        text (str): The input text to be cleaned.
+
+    Returns:
+        str: The cleaned text with special characters removed.
+    """
+
+    # Add a space between any digit and letter eg. 256GB -> 256 GB
+    text = re.sub(r"(\d+)([a-zA-Z])", r"\1 \2", text)
```
Comment (ShubhamVora-1206) [inline server/services/utils/processing.py:32]: Please create constants for regex patterns and import from there

**Pull request closed without merging**